indexed_gzip/zran_file_util.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * zran_file_util.c - file utilities used in zran.c to manipulate either
+ * Python file-like objects or file descriptors.
+ *
+ *
+ * Author: Paul McCarthy <pauldmccarthy@gmail.com>
+ */
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+#include "zran_file_util.h"
+
+#ifdef _WIN32
+#define FSEEK _fseeki64
+#define FTELL _ftelli64
+#include "windows.h"
+#include "io.h"
+#else
+#define FSEEK fseeko
+#define FTELL ftello
+#endif
+
+/*
+ * Implements a method analogous to fread that is performed on Python file-like objects.
+ */
+size_t _fread_python(void *ptr, size_t size, size_t nmemb, PyObject *f) {
+    PyObject *data;
+    char *buf;
+    Py_ssize_t len;
+    if ((data = PyObject_CallMethod(f, "read", "(n)", size * nmemb)) == NULL)
+        goto fail;
+    if ((buf = PyBytes_AsString(data)) == NULL)
+        goto fail;
+    if ((len = PyBytes_Size(data)) == -1)
+        goto fail;
+    memmove(ptr, buf, (size_t) len);
+    Py_DECREF(data);
+    return (size_t) len / size;
+
+fail:
+    Py_XDECREF(data);
+    return 0;
+}
+
+/*
+ * Implements a method analogous to ftell that is performed on Python file-like objects.
+ */
+int64_t _ftell_python(PyObject *f) {
+    PyObject *data;
+    uint64_t result;
+    if ((data = PyObject_CallMethod(f, "tell", NULL)) == NULL)
+        goto fail;
+    if ((result = PyLong_AsUnsignedLong(data)) == (unsigned long)-1 && PyErr_Occurred())
+        goto fail;
+    Py_DECREF(data);
+    return result;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to fseek that is performed on Python file-like objects.
+ */
+int _fseek_python(PyObject *f, int64_t offset, int whence) {
+    PyObject *data;
+    if ((data = PyObject_CallMethod(f, "seek", "(l,i)", offset, whence)) == NULL)
+        goto fail;
+    Py_DECREF(data);
+    return 0;
+
+fail:
+    Py_XDECREF(data);
+    return -1;
+}
+
+/*
+ * Implements a method analogous to feof that is performed on Python file-like objects.
+ * This method requires the file size to be input, because there is no other way to tell if a Python file-like is at EOF without reading it.
+ */
+int _feof_python(PyObject *f, int64_t size) {
```

Comment:
Why `int64_t` and not `uint64_t`?

Reply:
Good point, I changed it to `uint64_t` now.

Reply:
Actually, I had used `int64_t` earlier because `_ftell_python(f)` returns `int64_t`, and comparing a `int64_t` with `uint64_t` gives a warning:

![image](https://user-images.githubusercontent.com/1689183/110708757-29862980-81c9-11eb-81f6-3d84f58f9d6f.png)

For now, I just changed it to use a cast (`_ftell_python(f) == (int64_t) size;`), but let me know if you think of a cleaner solution.

Reply:
I suppose this can be revisited when somebody tries to load a 10PB gzip file!